Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Internal.Definitions;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.InputTypes;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
+using System.Collections;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Internal.Devices.WindowsMixedReality
+{
+    public struct WindowsMixedRealityController : IMixedRealityController<SixDoF>
+    {
+        public uint SourceId { get; set; }
+
+        public string SourceName { get; }
+
+        public IMixedRealityPointer[] Pointers { get; set; }
+
+        public InputType[] Capabilities { get; set; }
+
+        public Handedness Handedness { get; set; }
+
+        public SixDoF Controller { get; set; }
+
+        public SixDoF Pointer { get; set; }
+
+        public SixDoF Grip { get; set; }
+
+        public ControllerState ControllerState { get; set; }
+
+        public InteractionDefinition[] Interactions { get; set; }
+
+        public GameObject ControllerModel { get; set; }
+
+        //TODO - Needs re-implementing, as there may be more than left == right
+        #region IEquality Implementation
+
+        public static bool Equals(IMixedRealityInputSource left, IMixedRealityInputSource right)
+        {
+            return left.Equals(right);
+        }
+
+        public new bool Equals(object left, object right)
+        {
+            return left.Equals(right);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (ReferenceEquals(null, obj)) { return false; }
+            if (ReferenceEquals(this, obj)) { return true; }
+            if (obj.GetType() != GetType()) { return false; }
+
+            return Equals((IMixedRealityInputSource)obj);
+        }
+
+        private bool Equals(IMixedRealityInputSource other)
+        {
+            return other != null && SourceId == other.SourceId && string.Equals(SourceName, other.SourceName);
+        }
+
+        int IEqualityComparer.GetHashCode(object obj)
+        {
+            return obj.GetHashCode();
+        }
+
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                int hashCode = 0;
+                hashCode = (hashCode * 397) ^ (int)SourceId;
```

Comment:
where did the 397 come from?

Reply:
the number can/should be random. This number is used to salt the hash number.